### DISCUSSION TITLE
How to register design-time ViewModel

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8526] [Q&A]
Using this, you'd think it would register a design-time ViewModel.

Instead, it gives this error:

     ViewModelLocator.cs(50, 13): [SPLATDI006] MainViewModel has been registered in multiple places

What's the right way to do it?

```c#
public static class ViewModelLocator
{
    static ViewModelLocator()
    {
        var container = Locator.CurrentMutable;
            
        if (Design.IsDesignMode)
        {
            SplatRegistrations.Register<MainViewModel, MainViewModelDesign>();
        }
        else
        {
            SplatRegistrations.Register<MainViewModel>();
        }

        SplatRegistrations.SetupIOC();
    }

    public static MainViewModel Main => Locator.Current.GetService<MainViewModel>()!;

    public static void Cleanup()
    {
    }
}
```

**Comment by timunie** [comment]
I often use a static instance for design mode. I don't know if that would work in your case.

**Reply by mysteryx93** [reply]
well I do use a static

    d:DataContext="{x:Static local:ViewModelLocator.Main}"

but a static registered in the ViewModelLocator

I guess this would also be a solution, although that doesn't really answer the question as to why it's not working. Although doing this without the IOC doesn't allow me to feed in dependencies, so the ViewModel needs to be tweaked accordingly.

    d:DataContext="{x:Static local:ViewModelLocator.MainDesign}"

**Reply by maxkatz6** [reply]
I never used Splat as a dependency container, but Avalonia doesn't really care how static property was initialized.
Can you set breakpoint inside of Register property and see if it's registered more than once for MainViewModel? Or maybe there is a private list of current registrations.

**Reply by mysteryx93** [reply]
I cannot set breakpoints for design-time, but it does bring the question as to why both get registered.

Does the method get invoked twice? Or `SplatRegistrations.Register` injects IL at compile-time, does it inject it in a way that bypasses the conditional and both get called? It looks like the later.

**Reply by mysteryx93** [reply]
Sending this over to the Splat side
https://github.com/reactiveui/splat/issues/936

**Reply by maxkatz6** [reply]
You can replace "Design.IsDesignMode" with "true", so you can test locator.

**Reply by mysteryx93** [reply]
The dev says

> Yeah this is as designed since source generators are a compile time feature and it injects into a partial class. This warning is giving you a heads up.

**Comment by mysteryx93** [comment]
With a better understanding of the problem, this is the best I could do.

For most ViewModels, I do not like to create an interface because any change like adding a button or a label requires editing the ViewModel, the interface, and the ViewModelDesigner. For no real benefit.

This solution requires creating `IMainViewModel`... even though MainViewModelDesigner still inherits MainViewModel to avoid having to re-implement every member. `IMainViewModel` could also be an empty interface...

```c#
    SplatRegistrations.Register<MainViewModel>();
#if DEBUG
    SplatRegistrations.Register<MainViewModelDesign>();
    container.Register<IMainViewModel>(() => Design.IsDesignMode ? Locator.Current.GetService<MainViewModelDesign>() : Locator.Current.GetService<MainViewModel>());
#else
    container.Register<IMainViewModel>(() => Locator.Current.GetService<MainViewModel>());
#endif

...

public static IMainViewModel Main => Locator.Current.GetService<IMainViewModel>()!;
or
public static MainViewModel Main => (MainViewModel)Locator.Current.GetService<IMainViewModel>()!;
```

Got any idea to make this less ugly?

**Comment by mysteryx93** [comment]
Oh wait I found something better. No need for interface, using named contract instead. Here MainViewModelDesign just needs to have a parameterless constructor and take care of its own dependencies.

```c#
SplatRegistrations.Register<MainViewModel>("Init");
container.Register(() => Design.IsDesignMode ? new MainViewModelDesign() : Locator.Current.GetService<MainViewModel>("Init"));
```
